src/ngraph/distributed.cpp

```diff
@@ -14,37 +14,90 @@
 // limitations under the License.
 //*****************************************************************************
 
-#ifdef NGRAPH_DISTRIBUTED
+#ifdef NGRAPH_DISTRIBUTED_ENABLE
 
+#ifdef NGRAPH_DISTRIBUTED_MLSL_ENABLE
 #include <mlsl.hpp>
+#elif NGRAPH_DISTRIBUTED_OMPI_ENABLE
+#include <mpi.h>
+#endif
 
 #include "ngraph/distributed.hpp"
+#include "ngraph/log.hpp"
 
 using namespace ngraph;
 
 ngraph::Distributed::Distributed()
 {
+#ifdef NGRAPH_DISTRIBUTED_MLSL_ENABLE
     if (!MLSL::Environment::GetEnv().IsInitialized())
     {
         MLSL::Environment::GetEnv().Init(nullptr, nullptr);
+        this_init_comm = true;
+    }
+#elif NGRAPH_DISTRIBUTED_OMPI_ENABLE
+    int flag = 0;
+    MPI_Initialized(&flag);
+    if (!flag)
+    {
+        MPI_Init(NULL, NULL);
+        this_init_comm = true;
     }
+#else
+    throw ngraph_error("Distributed Library not supported/mentioned");
+#endif
 }
 
 ngraph::Distributed::~Distributed()
 {
+    if (this_init_comm == true)
+    {
+        finalize();
+    }
+}
+
+void ngraph::Distributed::finalize()
+{
+#ifdef NGRAPH_DISTRIBUTED_MLSL_ENABLE
     if (MLSL::Environment::GetEnv().IsInitialized())
```

Comment:
We could rely on `this_init_comm` flag and skip library specific check, e.g.
```
void ngraph::Distributed::finalize()
{
    if (this_init_comm)
    {
#ifdef NGRAPH_DISTRIBUTED_MLSL_ENABLE
        MLSL::Environment::GetEnv().Finalize();
#elif NGRAPH_DISTRIBUTED_OMPI_ENABLE
        MPI_Finalize();
#endif
    }
}
```

Reply:
I would like to leave this as is, for additional reliability :)